### ISSUE TITLE
Reading BLOB from SQLite3 no longer working

### ISSUE THREAD

**Issue by Tectu** [SOCI/soci#1173]
Hi,

I've updated an application using SOCI from `9bcc5f8a9181886f4c73ea5b4671b35d8722cb3a` to `a55c80984a447bf8c788a7d6a23ff93c022069fa`.

The only problem I encountered so far is with regards of reading a `BLOB` from an SQLite3 database.
Previously, the application did this (inside of a custom/user type conversion):
```
std::string str = v.get<std::string>("my_column");
```
Where:
- `my_column` is an SQLite3 column of type `BLOB`
- `v` is an lvalue reference to `soci::base_type`

This used to work well. However, after the update the code is throwing an `std::bad_cast` in `soci::details::holder::get()` (called by `soci::row::get()`).

According to the git log there have been some activities around blob support.

What is the new, correct way of reading (and writing!) a blob to/from SQLite3?

**Comment by vadz** [comment]
Thanks for reporting this, it's definitely a regression and should be fixed because I don't see anything wrong with this code.

I thought we had some testing code doing essentially the same thing, but I was clearly wrong. It would be great if you could please make a patch adding the (now) failing test case, this will help debug/bisect it and will ensure that it remains fixed in the future.

**Comment by Tectu** [comment]
Sure, I'll look into adding a test. Will take a while to get familiar with your testing setup tho.

I think the underlying "issue" or "cause" is that `soci::details::holder::get()` is clearly intentionally throwing when the type is `dt_blob`:
https://github.com/SOCI/soci/blob/a55c80984a447bf8c788a7d6a23ff93c022069fa/include/soci/type-holder.h#L338-L351

There's a comment indicating that this is due to blob being non-copyable.

**Comment by Krzmbrzl** [comment]
You can't obtain a BLOB as a string anymore. You need to do
```cpp
soci::blob blob = v.move_as<soci::blob>("my_column");
```

Relevant test case for this:
https://github.com/SOCI/soci/blob/c5678e172bff57c61c9a3a60aa253cebf98837ea/tests/common/test-lob.cpp#L643

Quoting from #992:
> When selecting a blob type via a rowset, they are now actually represented by a blob object instead of a std::string. The blob can be obtained from a given row via the new move_as<> function that complements the already existing get<> function for non-copyable types (such as blobs). Trying to use get<> for blob types will yield a compiler error stating that this is not supported.

**Comment by Tectu** [comment]
@Krzmbrzl Thank you for chiming in on this.

`move_as()` does not seem to be an available function when implementing a specialization of `type_conversion` (i.e. there is no `base_type::move_as()`).
What am I missing here?

**Comment by vadz** [comment]
> @vadz In order to restore backwards compatibility with the `.get<std::string>()` API, we could try to implement some mechanism to use the underlying `soci::blob` object to read its entire content into a string and then return that.

Yes, I'd like to do this, the code in the original code example looks fine to me and I think it's perfectly reasonable to represent (C)LOBs with `std::string` internally (even if `std::vector<byte>` would be better).

> However, for that to work we would have to make `.get<T>()` work when `T` and the underlying type are in principle incompatible (`std::string` and `soci::blob` in this case).

Do you understand how this worked before? 

> I don't think this can be done by means of a simple cast, can it? Also, we would likely want to emit some sort of deprecation warning as this seems to be less efficient (in general) that using the `soci::blob` object directly (as using it does not necessarily require reading the entire blob content into a string, which might become huge).

I don't think this warrants a warning, there are plenty of use cases when your BLOBs are not all that huge where this is fine. And anybody working with gigabyte-sized BLOBs should hopefully already understand that loading them in a string is not the brightest idea.

**Comment by Krzmbrzl** [comment]
@Tectu 

> move_as() does not seem to be an available function when implementing a specialization of type_conversion (i.e. there is no base_type::move_as()).
What am I missing here?

The `move_as` functions are direct member functions of the `row` class and are implemented in terms of `type_conversions<T>::move_as()`:
Note the template specialization for blobs though:
https://github.com/SOCI/soci/blob/c5678e172bff57c61c9a3a60aa253cebf98837ea/src/core/row.cpp#L112-L125

----

@vadz 

> Do you understand how this worked before?

I had something in mind about BLOBs being recognized as string types internally, but I don't think that is actually true... I'm going to look this up and see if I can write a fix for this issue.

**Comment by Krzmbrzl** [comment]
Turns out that this has been a lot more straight forward than I expected :D

**Comment by Tectu** [comment]
I can confirm that things are working as expected since #1189.

Thank you guys - much appreciated!
